Sources/NIOPerformanceTester/WebSocketFrameEncoderBenchmark.swift

```diff
@@ -0,0 +1,119 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import NIOWebSocket
+
+final class WebSocketFrameEncoderBenchmark {
+    private let channel: EmbeddedChannel
+    private let dataSize: Int
+    private let data: ByteBuffer
+    private let runCount: Int
+    private let dataStrategy: DataStrategy
+    private let cowStrategy: CoWStrategy
+    private var frame: WebSocketFrame?
+
+    init(dataSize: Int, runCount: Int, dataStrategy: DataStrategy, cowStrategy: CoWStrategy) {
+        self.channel = EmbeddedChannel()
+        self.dataSize = dataSize
+        self.runCount = runCount
+        self.dataStrategy = dataStrategy
+        self.cowStrategy = cowStrategy
+        self.data = ByteBufferAllocator().buffer(size: dataSize, dataStrategy: dataStrategy)
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark {
+    enum DataStrategy {
+        case spaceAtFront
+        case noSpaceAtFront
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark {
+    enum CoWStrategy {
+        case always
+        case never
+    }
+}
+
+
+extension WebSocketFrameEncoderBenchmark: Benchmark {
+    func setUp() throws {
+        // We want the pipeline walk to have some cost.
```

Comment:
I assume this is to make it closer to a "real" use case?

Reply:
exactly

Reply:
Yeah: the question this benchmark is going to try to answer is how useful it is to avoid an extra pipeline walk, so we need to have the pipeline walk have some actual overhead.